### PR TITLE
Omit module prefix when printing types which are reachable from Main

### PR DESCRIPTION
**Pull request by nalimilan**
Only add the module prefix when it is actually required to access the type from `Main`.

As discussed on Slack some time ago, this is an attempt at improving the printing of types defined outside Base. I'm really not familiar with this, so there might be better ways of doing this. Also, I'm not sure how to test this given that tests do not run in Main.

The improvement is the most visible for types defined in packages for which the package name is just the plural of the type name. Before:
```julia
julia> using CategoricalArrays

julia> CategoricalArray(1:2)
2-element CategoricalArrays.CategoricalArray{Int64,1,UInt32}:
...
```

After:
```julia
julia> using CategoricalArrays

julia> CategoricalArray(1:2)
2-element CategoricalArray{Int64,1,UInt32}:
...
```

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
I don't think that making `Main` more special here is the right way to go. Instead, this should be handled by doing `IOContext(io, :module => Main)` from the REPL and then adding generic logic to only print as much as is required to reconstruct a name with respect to the `get(io, :module, Empty)` where `const Empty = Module(:Empty)`. I also think that searches for aliases could be done per context module so that we know how to print a short version of an object in some context.

**Comment by nalimilan** [comment]
Good idea. I've updated the PR to add an `IOContext` mechanism and used it for `Main` from the REPL.

> with respect to the get(io, :module, Empty) where const Empty = Module(:Empty).

I've tried using that dummy module as a default, but that prints lots of warnings like "importing deprecated binding Base.airy into Empty" when calling `sprint(dump, Any)` (which happens in a test). I don't understand why this happens, so I've used `get(io, :module, nothing)` instead.

> I also think that searches for aliases could be done per context module so that we know how to print a short version of an object in the context of some module, which will often be Main but not always.

What do you mean by "context module"? In general it's hard to know what's the relevant module, as the string generated by `show` can be passed to different functions and eventually be shown to the user in a very different context.

File: test/repl.jl [hunk 1]
```diff
@@ -167,6 +167,21 @@ fake_repl() do stdin_write, stdout_read, repl
     close(t)
     readuntil(stdout_read, "\n\n")
 
+    # Test that module prefix is omitted when type is reachable from Main
```
Comment (nalimilan) [inline test/repl.jl]: I couldn't find a simpler way to test `display(::REPLDisplay, ...)`. If there's a cleaner solution, I'm all ears.

Comment (yuyichao) [reply test/repl.jl]: Well, as I said, you can

> eval [...] in `Main`

Comment (nalimilan) [reply test/repl.jl]: But eval what? :-)

Comment (yuyichao) [reply test/repl.jl]: Whatever you need to be run in `Main`?

Instead of `write(stdin_write, "module TestShowTypeREPL; export TypeA; struct TypeA end; end\n")` and `write(stdin_write, "using .TestShowTypeREPL\n")` just do `eval` instead?

`@eval Main module TestShowTypeREPL; export TypeA; struct TypeA end; end` and `@eval Main using .TestShowTypeREPL`

**Comment by StefanKarpinski** [comment]
Using `get(io, :module, nothing)` is probably better anyway!

> What do you mean by "context module"?

I just meant whatever the value of `get(io, :module)` is.

**Comment by KristofferC** [comment]
Cross reffing https://github.com/JuliaDocs/Documenter.jl/pull/574#issuecomment-335038914 since it is a bit related.

**Comment by KristofferC** [comment]
Another cross ref: https://github.com/JuliaLang/julia/issues/24059.

File: test/repl.jl (hunk 1 above)
Comment (nalimilan) [reply test/repl.jl]: Thanks, that makes the code simpler. Though I haven't found a way to get the result without the whole `readline` stuff.

File: base/show.jl [hunk 2]
```diff
@@ -1992,7 +2007,7 @@ function showarray(io::IO, X::AbstractArray, repr::Bool = true; header = true)
         # override usual show method for Vector{Method}: don't abbreviate long lists
         io = IOContext(io, :limit => false)
     end
-    (!repr && header) && print(io, summary(X))
```
Comment (rfourquet) [inline base/show.jl]: This change seems unrelated?

Comment (nalimilan) [reply base/show.jl]: Right, but probably not worth a PR?

Comment (nalimilan) [reply base/show.jl]: Oh, and it's not completely unrelated since it's needed so that the module context is passed to `summary` via `io`. So it's useful on its own, but needed for this PR.

File: base/show.jl [hunk 3]
```diff
@@ -241,6 +241,14 @@ end
 
 show(io::IO, x::DataType) = show_datatype(io, x)
 
+# Check whether 'sym' (defined in module 'parent') is reachable from module 'from'
+# If an object with this name exists in 'from', we need to check it's the same object
+function is_reachable(sym::Symbol, parent::Module, from::Module)
+    isdefined(from, sym) || return false
+    getfield(from, sym) === getfield(parent, sym)
+end
+is_reachable(sym::Symbol, parent::Module, from::Void) = false
```
Comment (rfourquet) [inline base/show.jl]: as `isdefined(nothing, :anysymbol) == false`, the 2 methods could be simplified to `is_reachable(sym::Symbol, parent::Module, from::Union{Module,Void}) = isdefined(from, sym) && getfield(from, sym) === getfield(parent, sym)`. Not saying it's necessarily an improvement!

Comment (nalimilan) [reply base/show.jl]: Yeah, at least the current definition is quite clear, if a bit verbose.

**Comment by nalimilan** [comment]
Bump. I think this is good for a review.

**Comment by StefanKarpinski** [comment]
I will review soon!

File: base/repl/REPL.jl [hunk 4]
```diff
@@ -122,7 +122,7 @@ end
 function display(d::REPLDisplay, mime::MIME"text/plain", x)
     io = outstream(d.repl)
     Base.have_color && write(io, answer_color(d.repl))
-    show(IOContext(io, :limit => true), mime, x)
+    show(IOContext(io, :limit => true, :module => Main), mime, x)
```
Comment (StefanKarpinski) [inline base/repl/REPL.jl]: We may want to allow giving the module via a keyword, but that can be added later.

File: base/show.jl (hunk 3 above)
Comment (StefanKarpinski) [reply base/show.jl]: I find @rfourquet's formulation clearer, but people's brains are different, so YMMV. This isn't a public function so the name is not crucial, but this might be better called `isvisible`; `isreachable` sounds like there's some kind of recursive search going on, which there isn't – it's just an immediate visibility check.

File: base/show.jl [hunk 5]
```diff
@@ -266,7 +274,14 @@ function show_type_name(io::IO, tn::TypeName)
     elseif globfunc
         print(io, "typeof(")
     end
-    if isdefined(tn, :module) && !(is_exported_from_stdlib(sym, tn.module) || (tn.module === Main && !hidden))
+    # Print module prefix unless type is:
```
Comment (StefanKarpinski) [inline base/show.jl]: Wasn't immediately clear to me without reading the code if this was a conjunction or disjunction of conditions, so maybe change the comment to this:
```jl
# Print module prefix unless type satisfies one of the following:
```
But see comment below which would change this to just check visibility.

File: base/show.jl [hunk 6]
```diff
@@ -266,7 +274,14 @@ function show_type_name(io::IO, tn::TypeName)
...
+    # - exported from standard library
+    # - reachable from module passed to IOContext
+    # - defined in Main and not hidden
+    if isdefined(tn, :module) &&
+        !(is_exported_from_stdlib(sym, tn.module) ||
+          is_reachable(sym, tn.module, get(io, :module, nothing)) ||
+          (tn.module === Main && !hidden))
```
Comment (StefanKarpinski) [inline base/show.jl]: I feel like it might be better to *only* consider visibility. Doesn't that cover the other cases in the typical situations where we are displaying from `Main` and the `Base` has been imported? If the user asks to display something with respect to another module, wouldn't we want to only shorten names that are actually visible in that module? Why are `Main` or stdlib relevant there?

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]
Generally good, my major comment is about the criteria for when to print prefixes. I think it we should *only* consider visibility with respect to the given module. If no module is given, always print a fully qualified name.

File: base/show.jl (hunk 6 above)
Comment (nalimilan) [reply base/show.jl]: I've just tried it, and it triggers a lot of test failures all over the place. That made me realize the `IOContext` approach implies hardcoding `:module => Main` in many places. For example, `stringmime` calls `_textreprmime(m::MIME, x) = sprint(verbose_show, m, x)`. To omit the module prefix, we need to create the `IOBuffer` by hand to pass it an adequate `IOContext`. There are many examples like that.

Do we really want to add this kind of thing all over the place? I'm asking before spending time to find them all. The problem is that in the end many functions do not take a `IO` object, so they will hardcode `Main`, which kind of defeats the use of `IOContext`. Is that still worth it, rather than hardcoding `Main` in a single central place?

Comment (nalimilan) [reply base/show.jl]: Actually, a good intermediate solution would be to default to `Main` (i.e. use `get(io, :module, Main)` instead of `get(io, :module, nothing)`), but still allow passing a different `:module`. That way by default we get reasonably short type names without hardcoding `Main` in many places, but it can still be customized.

Comment (StefanKarpinski) [reply base/show.jl]: That sounds like a good approach: `Main` as the default `nothing` as a value that can be passed to have no context module. Another thing that occurs to me is allowing an array of modules to look in and having `[]` indicate not to look anywhere, defaulting to `[Main]`. But it would also make sense to allow a single module as a shortcut in that case. Maybe too complicated as a starting point.

**Comment by nalimilan** [comment]
I've implemented the new approach which falls back to `Main` by default.

@StefanKarpinski You haven't commented on the fact that functions are always printed with the full prefix (i.e. not change). Does that sound reasonable to you? That's kind of inconsistent, but in practice I think you're generally interested in knowing where a function is define, notably to add methods to it. But I could be convinced otherwise.

**Comment by StefanKarpinski** [comment]
I do think that knowing where a function comes from is useful.

File: base/show.jl [hunk 7]
```diff
@@ -266,7 +273,9 @@ function show_type_name(io::IO, tn::TypeName)
     elseif globfunc
         print(io, "typeof(")
     end
-    if isdefined(tn, :module) && !(is_exported_from_stdlib(sym, tn.module) || (tn.module === Main && !hidden))
+    # Print module prefix unless type is visible from module passed to IOContext
+    # (defaulting to Main)
+    if isdefined(tn, :module) && !isvisible(sym, tn.module, get(io, :module, Main))
```
Comment (StefanKarpinski) [inline base/show.jl]: What about pulling out `get(io, :module, Main)` and checking if it's `nothing` instead of adding the `Void` method to `isvisible`? That method seems a bit iffy to me.

Comment (nalimilan) [reply base/show.jl]: Done.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by fredrikekre** [comment]
I agree with Alex here: https://github.com/JuliaLang/julia/pull/25798#issuecomment-361367528, I liked the idea, but after some experience with it I don't really like it anymore. What do other people think? I agree that printing `CategoricalArrays` in `CategoricalArrays.CategoricalArray` from OP seem rather superfluous when the type directly matches the package name like that, but consider e.g.
```julia
julia> A'
2×2 LinearAlgebra.Adjoint{Float64,Array{Float64,2}}:
 0.628495  0.0341545
 0.693558  0.787898
```
and
```julia
julia> A'
2×2 Adjoint{Float64,Array{Float64,2}}:
 0.742562  0.773093
 0.875984  0.206749
```
Here I think it is informative to see the `LinearAlgebra.` prefix. Perhaps it would be better if this change was opt-in somehow. Thoughts?

(Unrelated, but seeing `LinearAlgebra.Adjoint` looks so much nicer than `Base.LinAlg.Adjoint` :smile: )

**Comment by quinnj** [comment]
Personally, I've really liked the new printing because it provides that extra piece of information of: what you see is what you can access. I.e. if it's prequalified w/ `LinearAlgebra.Adjoint`, then I know to access that type, I need to do `LinearAlgebra.Adjoint`. Before, we never had this information; you had to keep a mental tally of, "oh, even though it prints like `Base.LinearAlgebra.Adjoint`, I did `using LinearAlgebra`, so I don't need to fully qualify the type in my own code".

**Comment by StefanKarpinski** [comment]
I've found this to be incredibly pleasant to use and consider it a big success at making the way we print things more concise and intuitive. Note that I've added `repr(obj, :module => nothing)` and `repr(obj, :module => Base)` and other `IOContext` argument pairs so that if you want to know how to print something with respect to another module, it's easy enough. We do probably want to figure out a scheme for configuring the `IOContext` that the REPL uses for printing. That would allow doing something like `REPL.context[:module] = Base` to make the REPL print from the context of `Base` instead of `Main` or `REPL.context[:module] = nothing` to make it print every name in full. That would also allow `REPL.context[:float_digits] = 6` or something like that to control how significant many digits floating-point numbers are printed to.

**Comment by stevengj** [comment]
This PR gives a way to force the printing of a module (by setting `:module` to `nothing`), but not a way to forcibly suppress it.

I would suggest instead using `IOContext(io, :module => true)` to force printing of the module and `IOContext(io, :module => false)` to force it to be suppressed.

**Comment by nalimilan** [comment]
`:module => false` makes sense to me. Though you can pass the result of `parentmodule` to get the same result, at least when printing a single type.

Using `:compact => true` to drop type parameters is an interesting idea, but it could be misleading as it really removes information and it could be more problematic than rounding numbers.

**Comment by stevengj** [comment]
You could print Foo{...} to indicate that parameters are elided.

**Comment by stevengj** [comment]
Passing parentmodule is a pain when printing multiple types, and impossible for nested type parameters
